### ISSUE TITLE
Swift 5.2 compiler optimization crash

### ISSUE THREAD

**Issue by jstart** [postmates/PMKVObserver#26]
I've been putting off filing this issue because of procrastination and because I haven't been able to create a sample project to recreate the crash, but I'll try and provide as much info as possible.

I work for an unnamed food delivery service that rhymes with "roast dates". We recently updated to Swift 5.2 and found a crash in production that only happened in a certain scenario with compiler optimizations on. I found that you had already discovered some strange behavior while compiler optimizations are enabled here: https://github.com/postmates/PMKVObserver/commit/778390212076933d66325756a3f91d26698e4e34

So I updated the SDK, but still saw the same crash. The stack trace in production looks like this:
![Screen Shot 2020-05-26 at 4 25 45 PM](https://user-images.githubusercontent.com/108218/82959947-e591a300-9f6d-11ea-865b-1b084b889e7e.png)

And our code that is using the library looks like:
```
extension UILabel {
    private static var map = AssocMap<(style: TextStyle, observer: KVObserver)>()

    var textStyle: TextStyle? {
        get {
            return UILabel.map[self]?.style
        }

        set {
            if let style = newValue {
                let observer = KVObserver(object: self, keyPath: \UILabel.text, options: .new) { object, _, _ in
                    object.updateTextWithStyle()
                }

                UILabel.map[self] = (style: style, observer: observer)
                updateTextWithStyle() // Update last, once the token/observer and style are registered
            } else {
                UILabel.map[self]?.observer.cancel()
                UILabel.map[self] = nil
            }
        }
    }
```

We are using the `AssocMap` library as well. Do you think TSAO needs to be modified in a similar way as you did to this project?: https://github.com/lilyball/swift-tsao/blob/e15c51277b84be62cf003cd654ad02245f87e15d/Sources/TSAO.swift#L31

Why I hesitated to file this is because I'm not sure what `exactly` causes the crash.  The crash logs in Xcode would open unrelated places in the code where we were setting `.text` which makes me think it's accessing other labels in the AssocMap and finding deallocated labels. Let me know if I can provide any more info or if I should just wait until this Swift PR is merged and released and see if that resolves this issue: https://github.com/apple/swift/pull/31662

Thank you!

**Comment by lilyball** [comment]
The type-check in TSAO is a bit different, it's testing if the type conforms to `AnyObject`, whereas the Swift bug worked around in 778390212076933d66325756a3f91d26698e4e34 was about protocol types (e.g. `T.Protocol`). While I haven't looked at TSAO in a while, I doubt it's hitting the same issue. If it did, that would just result in unnecessary boxing. And in a quick test, what TSAO is doing works just fine in Swift 5.2.4.

> which makes me think it's accessing other labels in the AssocMap and finding deallocated labels

This can't be the case as `AssocMap` doesn't actually store any data at all, it's a wrapper around associated types, so `UILabel.map[self]` is actually a wrapper around `objc_getAssociatedValue(self, …)`.

What I'm curious about is you're reporting a crash with `UILabel.text.setter`, but that's an SDK method, what you've got in your code snippet is a separate property `textStyle`. And in particular, the crash you screenshotted is in `function signature specialization <Arg[0] = Exploded> of Swift.String._bridgeToObjectiveCImpl() -> Swift.AnyObject`. Which is to say, it's crashing when you assign a `String` to `UILabel.text` and it tries to bridge it to `NSString`. This suggests you've got some sort of damaged string object. I assume you're looking at PMKVObserver because you're observing the `text` property, except this crash appears to happen prior to even entering the SDK's actual `-[UILabel setText:]` method, much less invoking the KVO observer.

**Comment by jstart** [comment]
Thanks so much for the response! That makes a lot more sense. I just assumed it was related to this since I saw your commit fixing the issue around compiler optimization. We resolved the issue by rolling back to Swift 5.0. I'll do some more digging to see why the string might be damaged before being bridged. 👍

**Comment by jstart** [comment]
Closing this as I'm 90% sure the crash is totally unrelated to this library. Thanks so much for your help!!
